go/libkb/secret_store_darwin.go

```diff
@@ -7,138 +7,225 @@ package libkb
 	query := keychain.NewItem()
 	query.SetSecClass(keychain.SecClassGenericPassword)
-	query.SetService(k.serviceName(m))
-	query.SetAccount(accountName)
+	query.SetService(k.serviceName(mctx))
+	query.SetAccount(account.String())
 
 	// iOS keychain returns `keychain.ErrorParam` if this is set so we skip it.
 	if !isIOS {
 		query.SetMatchLimit(keychain.MatchLimitOne)
 	}
+
 	updateItem := keychain.NewItem()
 	updateItem.SetAccessible(k.accessible())
 	if err := keychain.UpdateItem(query, updateItem); err != nil {
-		m.Debug("KeychainSecretStore.updateAccessibility: failed: %s", err)
+		mctx.Debug("KeychainSecretStore.updateAccessibility: failed: %s", err)
 	}
 }
 
-func (k KeychainSecretStore) mobileKeychainPermissionDeniedCheck(m MetaContext, err error) {
-	m.G().Log.Debug("mobileKeychainPermissionDeniedCheck: checking for mobile permission denied")
-	if !isIOS || m.G().GetAppType() != MobileAppType {
-		m.G().Log.Debug("mobileKeychainPermissionDeniedCheck: not an iOS app")
+func (k KeychainSecretStore) mobileKeychainPermissionDeniedCheck(mctx MetaContext, err error) {
+	mctx.G().Log.Debug("mobileKeychainPermissionDeniedCheck: checking for mobile permission denied")
+	if !isIOS || mctx.G().IsMobileAppType() {
+		mctx.G().Log.Debug("mobileKeychainPermissionDeniedCheck: not an iOS app")
 		return
 	}
 	if err != keychain.ErrorInteractionNotAllowed {
-		m.G().Log.Debug("mobileKeychainPermissionDeniedCheck: wrong kind of error: %s", err)
+		mctx.G().Log.Debug("mobileKeychainPermissionDeniedCheck: wrong kind of error: %s", err)
 		return
 	}
-	m.G().Log.Warning("mobileKeychainPermissionDeniedCheck: keychain permission denied: %s", err)
+	mctx.G().Log.Warning("mobileKeychainPermissionDeniedCheck: keychain permission denied, aborting: %s", err)
 	os.Exit(4)
 }
 
-func (k KeychainSecretStore) RetrieveSecret(m MetaContext, accountName NormalizedUsername) (LKSecFullSecret, error) {
-	m.Debug("KeychainSecretStore.RetrieveSecret(%s)", accountName)
-	encodedSecret, err := keychain.GetGenericPassword(k.serviceName(m), string(accountName), "", "")
+func (k KeychainSecretStore) RetrieveSecret(mctx MetaContext, accountName NormalizedUsername) (secret LKSecFullSecret, err error) {
+	defer mctx.TraceTimed(fmt.Sprintf("KeychainSecretStore.RetrieveSecret(%s)", accountName), func() error { return err })()
+
+	// find the last valid item we have stored in the keychain
+	var previousSecret LKSecFullSecret
+	for i := 0; i < maxKeychainItemSlots; i++ {
+		account := newKeychainSlottedAccount(accountName, i)
+		secret, err = k.retrieveSecret(mctx, account)
+		if err == nil {
+			previousSecret = secret
+			mctx.Debug("successfully retrieved secret on attempt: %d, checking if there is another filled slot", i)
+		} else if _, ok := err.(SecretStoreError); ok || err == keychain.ErrorItemNotFound {
+			// We've reached the end of the keychain entries so let's return
+			// the previous secret we found.
+			secret = previousSecret
+			k.updateAccessibility(mctx, account)
```

Comment:
@joshblum do you remember why do we try to update an item that we know doesn't exist? or is that we are correcting for the case where item exists but is somehow inaccessible and we are hoping for it to work during next try?

Reply:
oh but that not what `updateAccessibility` is doing... I don't think this is right what's in here

Reply:
i don't remember why, it might have been keeping the previous functionality for this case. is it causing a specific issue?

Reply:
while I'm looking for issues in `secret_store_darwin.go`, i don't think this causes any of the issues I'm trying to solve.

I think `updateAccessibility` was used to change some parameter of existing keychain entries. I think `k.updateAccessibility(mctx, account)` should happen e.g. on line 152 - if an entry is found, use the opportunity to update it. At least that's what I gather is how it used to work becuase the change to multiple slots for username.

There may be a possibility that this code is not needed anymore, too.

I think @mmaxim wrote the original update code, maybe he will remember.